### PR TITLE
172915- Encrypt ENC key using Unencrypted Hardware Id

### PR DESCRIPTION
**Pull request by shirinbt14926**
This PR consists code changes for below PBIs

[PBI 172916](https://dev.azure.com/ukhydro/S-100%20Permit%20Generator/_workitems/edit/172916)
Code changes added
Unit Test cases added
Xsd validation and file added

[PBI 172915](https://dev.azure.com/ukhydro/S-100%20Permit%20Generator/_sprints/taskboard/S-100%20Permit%20Service/S-100%20Permit%20Generator/Sprint%205%20-%20WP%20BA2%20S-100%20Permit%20Service?workitem=172915)
Coding
Unit Tests

### CODE REVIEW

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs [hunk 1]
```diff
@@ -117,35 +137,93 @@ private void CreatePermitXml(DateTimeOffset issueDate, string dataServerIdentifi
...
+
+        private List<ProductKeyServiceRequest> ProductKeyServiceRequest(
             IEnumerable<HoldingsServiceResponse> holdingsServiceResponse) =>
             holdingsServiceResponse.SelectMany(x => x.Cells.Select(y => new ProductKeyServiceRequest
             {
                 ProductName = y.CellCode,
                 Edition = y.LatestEditionNumber
             })).ToList();
+
+        private string EncryptKey(IEnumerable<ProductKey> productKey, string hardwareId, HoldingsServiceResponse holding) =>
```
Comment (Nadeem14456) [inline src/UKHO.S100PermitService.Common/Services/PermitService.cs:224]:
```suggestion
        private string GetEncryptedKey(IEnumerable<ProductKey> productKeys, string hardwareId, HoldingsServiceResponse holdings) =>
```

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs [hunk 2]
```diff
@@ -117,35 +137,93 @@ private void CreatePermitXml(DateTimeOffset issueDate, string dataServerIdentifi
...
+            _s100Crypt.CreateEncryptedKey((from str1 in holding.Cells
+                                         join str2 in productKey on str1.CellCode.ToString() equals str2.ProductName
+                                         select str2.DecryptedKey).FirstOrDefault(), hardwareId);
```
Comment (Nadeem14456) [inline src/UKHO.S100PermitService.Common/Services/PermitService.cs:227]: can split this for better readability. First retrieve the productKey then make a call to _s100Crypt.CreateEncryptedKey.
```suggestion
            var productKey = holding.Cells
    .Join(productKeys,
        cell => cell.CellCode.ToString(),
        key => key.ProductName,
        (cell, key) => key.DecryptedKey)
    .FirstOrDefault();

return _s100Crypt.CreateEncryptedKey(productKey, hardwareId);
```

File: tests/UKHO.S100PermitService.Common.UnitTests/Services/PermitServiceTests.cs
Comment (Nadeem14456) [inline tests/UKHO.S100PermitService.Common.UnitTests/Services/PermitServiceTests.cs:1]: A call to below method must be covered in permit service?

_s100Crypt.CreateEncryptedKey(productKey, hardwareId);

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]

File: src/UKHO.S100PermitService.Common/Encryption/S100Crypt.cs [hunk 3]
```diff
@@ -39,7 +41,7 @@ public IEnumerable<ProductKey> GetDecryptedKeysFromProductKeys(IEnumerable<Produ
                 });
             }
 
-            _logger.LogInformation(EventIds.GetDecryptedKeysFromProductKeysCompleted.ToEventId(), "Get decrypted keys from product keys completed.");
+            _logger.LogInformation(EventIds.GetDecryptedKeysFromProductKeysCompleted.ToEventId(), "Get decrypted keys from Product keys completed.");
```
Comment (harshal11869) [inline src/UKHO.S100PermitService.Common/Encryption/S100Crypt.cs:44]: To keep it consistent as line no. 30 no need to change
```suggestion
            _logger.LogInformation(EventIds.GetDecryptedKeysFromProductKeysCompleted.ToEventId(), "Get decrypted keys from product keys completed.");
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: src/UKHO.S100PermitService.Common/Encryption/S100Crypt.cs (hunk 3 above)
Comment (shirinbt14926) [reply src/UKHO.S100PermitService.Common/Encryption/S100Crypt.cs:44]: done

File: tests/UKHO.S100PermitService.Common.UnitTests/Services/PermitServiceTests.cs
Comment (shirinbt14926) [reply tests/UKHO.S100PermitService.Common.UnitTests/Services/PermitServiceTests.cs:1]: added

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs (hunk 2 above)
Comment (shirinbt14926) [reply src/UKHO.S100PermitService.Common/Services/PermitService.cs]: Refactored as suggested

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs (hunk 1 above)
Comment (shirinbt14926) [reply src/UKHO.S100PermitService.Common/Services/PermitService.cs]: done

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs [hunk 4]
```diff
@@ -117,35 +137,93 @@ private void CreatePermitXml(DateTimeOffset issueDate, string dataServerIdentifi
         }
 
         [ExcludeFromCodeCoverage]
-        private static List<Products> GetProductsList()
+        private List<Products> GetProductsList(List<HoldingsServiceResponse> holdings, IEnumerable<ProductKey> productKey, string hardwareId, string UpnTitle)
```
[the remark this answers is not in the document]
Comment (shirinbt14926) [reply src/UKHO.S100PermitService.Common/Services/PermitService.cs]: done

File: src/UKHO.S100PermitService.Common/Encryption/S100Crypt.cs
[the remark this answers is not in the document]
Comment (shirinbt14926) [reply src/UKHO.S100PermitService.Common/Encryption/S100Crypt.cs:1]: Removed

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shirinbt14926) [reply src/UKHO.S100PermitService.Common/Services/PermitService.cs]: updated

**Review by harshal11869 — APPROVED** [review]
LGTM.

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs [hunk 5]
```diff
@@ -117,35 +137,98 @@ private void CreatePermitXml(DateTimeOffset issueDate, string dataServerIdentifi
         }
 
         [ExcludeFromCodeCoverage]
-        private static List<Products> GetProductsList()
+        private List<Products> GetProductsList(List<HoldingsServiceResponse> holdingsServiceResponse, IEnumerable<ProductKey> decryptedProductKeys, string hardwareId, string upnTitle)
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Services/PermitService.cs:140]: Both list parameters here should be IEnumerable<>

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs [hunk 6]
```diff
@@ -137,7 +137,7 @@ private void CreatePermitXml(string userPermit, string title, List<Products> pro
         }
 
         [ExcludeFromCodeCoverage]
-        private List<Products> GetProductsList(List<HoldingsServiceResponse> holdingsServiceResponse, IEnumerable<ProductKey> decryptedProductKeys, string hardwareId, string upnTitle)
+        private List<Products> GetProductsList(IEnumerable<HoldingsServiceResponse> holdingsServiceResponse, IEnumerable<ProductKey> decryptedProductKeys, string hardwareId, string upnTitle)
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Services/PermitService.cs:140]: The method should return an IEnumerable<Products> as well

**Review by jrippington — CHANGES_REQUESTED** [review]
Still not quite right

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs (hunk 6 above)
Comment (shirinbt14926) [reply src/UKHO.S100PermitService.Common/Services/PermitService.cs]: Resolved

**Review by jrippington — APPROVED** [review]
Looks good, thanks!

**Pull request merged**